src/realm/sync/subscriptions.hpp

```diff
@@ -0,0 +1,290 @@
+                o << "Uncommitted";
+                break;
+            case State::Pending:
+                o << "Pending";
+                break;
+            case State::Bootstrapping:
+                o << "Bootstrapping";
+                break;
+            case State::Complete:
+                o << "Complete";
+                break;
+            case State::Error:
+                o << "Error";
+                break;
+        }
+        return o;
+    }
+
+    class iterator {
+    public:
+        using difference_type = size_t;
+        using value_type = Subscription;
+        using pointer = value_type*;
+        using reference = value_type&;
+        using iterator_category = std::input_iterator_tag;
+
+        bool operator!=(const iterator& other) const
+        {
+            return (m_parent != other.m_parent) || (m_sub_it != other.m_sub_it);
+        }
+
+        bool operator==(const iterator& other) const
+        {
+            return (m_parent == other.m_parent && m_sub_it == other.m_sub_it);
+        }
+
+        reference operator*() const noexcept
+        {
+            return m_cur_sub;
+        }
+
+        pointer operator->() const noexcept
+        {
+            return &m_cur_sub;
+        }
+
+        // used in tests.
+        inline friend std::ostream& operator<<(std::ostream& o, const iterator& it)
+        {
+            o << "SubscriptionSet::iterator(" << std::hex << it.m_parent << ", " << std::dec << it.m_sub_it.index()
+              << ")";
+            return o;
+        }
+
+        iterator& operator++();
+        iterator operator++(int);
+
+    protected:
+        friend class SubscriptionSet;
+
+        iterator(const SubscriptionSet* parent, LnkLst::iterator it);
+
+    private:
+        const SubscriptionSet* m_parent;
+        LnkLst::iterator m_sub_it;
+        mutable Subscription m_cur_sub;
+    };
+
+    using const_iterator = const iterator;
+
+    // This will make a copy of this subscription set with the next available version number and return it as
+    // a mutable SubscriptionSet to be updated. The new SubscriptionSet's state will be Uncommitted. This
+    // subscription set will be unchanged.
+    SubscriptionSet make_mutable_copy() const;
+
+    // The query version number used in the sync wire protocol to identify this subscription set to the server.
+    int64_t version() const;
+
+    // The current state of this subscription set
+    State state() const;
+
+    // The error string for this subscription set if any.
+    StringData error_str() const;
+
+    // Returns the number of subscriptions in the set.
+    size_t size() const;
+
+    // A const_iterator interface for finding/working with individual subscriptions.
+    const_iterator begin() const;
+    const_iterator end() const;
+
+    // Returns a const_iterator to the query matching either the name or Query object, or end() if no such
+    // subscription exists.
+    const_iterator find(StringData name) const;
```

Comment:
I am apprehensive of using the query name like this. Can we just use an UUID or ObjectId for identification? 

1. How would we handle a server response to update the query state if this query has changed its name locally in the meantime?
2. This ties the query description to the FLX metadata format such that if the description changes between Realm releases this might break.
3. There are multiple strings that could describe the same underlying Query and they might not be the one that the user first writes (eg: `name == 'foo'` == `!('foo' != name)`)

Reply:
The purpose of these names is to give users the ability to set a convenient custom name for queries. So you could have a query named "orders_page" that represents the query that drives your orders page UI, and you want to be able to adjust it by saying `realm.get_latest_subscriptions().update([&](SubscriptionSet* set) { set->insert_or_assign("orders_page", new_query});` The names are never sent to the server, they're just for the client to be able to easily find queries they've already created.

You can also have anonymous queries that have no name. For them, with the current API, you'd just iterate all the subscriptions and filter the ones that match your table. Are you basically asking for another identifier type to uniquely identify each subscription that you get even if you don't choose a custom name?

As for handling a server response, the server is dealing with whole subscription sets. All the queries on the same table get or'd together before getting sent to the server. The SubscriptionSet class is designed so that if you want to modify a query, you have to make a copy of it to a new SubscriptionSet with its own version number. The version number is how the server will identify the SubscriptionSet when it sends any state or errors about the subscription set.

I think the second two points are things I'm apprehensive about as well, but I'm not sure what the best way to mitigate the risk is. Do you think this risk is big enough that we cannot move forward with an MVP for SDKs until we address it?

Reply:
Treating all queries in a SubscriptionSet atomically and using the version number as the identifier solves most of my concerns about how this will be used. Not having to handle updating individual queries simplifies everything. I guess there are still some edge cases around when the user tries to name multiple queries the same thing where having an id would still be useful instead of using indexes or something. This is certainly good enough for MVP though and not blocking.

Unrelated question: the description we will be sending to the server is going to be a MongoDB formatted query string generated by the Query class on the client side correct?

Reply:
Actually, it's going to be sent on the wire as Realm query strings and the server will transform it into MQL on its end.

Reply:
Interesting, I can see the benefit of that, but it also means the server team has to write another query parser for core's query strings.